test/tls-provider.c

```diff
@@ -15,6 +15,11 @@
 /* For TLS1_3_VERSION */
 #include <openssl/ssl.h>
 
+static OSSL_FUNC_keymgmt_import_fn xor_import;
```

Comment:
Does the addition mean that with no-cache providers you have to always provide export/import functions and that there would be import/export done on all operations with a key? Also the commit message for this commit is confusing me. Should the last sentence be `Without allowing import and export, an existing key can not move to a new key manager even if it is really the same.`?

Reply:
Currently, I think they are necessary.  @levitte suggested that this wasn't the intention on Tuesday, so it might be fixable.

I'll fix the commit message, you are correct.

Reply:
Thinking about the typical case of the no_cache provider being the PKCS#11 provider - is it indication of some inherent issue with the concept? The problem is that sometimes the import/export would actually be needed - what if the token that handled the original key is replaced? But on the other hand in many cases the import/export actually won't be possible at all - thinking about unexportable private keys residing inside the token.

Reply:
I don't think the problem is widespread.  Only one test case failed.  This was at the call to `create_ssl_connection()` in the `test_pluggable_group` test in `test/sspapitest.c`.  It is triggered when an algorithm is fetched to do something with an existing key.  Even though the algorithm is the same, the lack of caching means it is different and thus the export is required.

I'll note that this problem is **not** restricted to `no_cache`.  Anything that causes the fetch cache to be flushed would trigger exactly the same thing.  In normal use, I believe that this would be quite unlikely.  It would also be **very** difficult to debug.  All of the problems I've identified with `no_cache` are also present in normal builds.

I'm expecting the PKCS#11 provider to uncover a multitude of problems.  We will need the `no_cache` behaviour for it, we likely will need more changes.

Reply:
The import/export dance is necessary when the operation provider and the keymgmt provider are different.  If this happens when the two providers are the same, then we're doing something stoopid when comparing, or we have loaded the same provider twice.

Reply:
Do note that loading the same provider twice is perfectly legitimate.  They may be held in different library contexts.  That's a perfect way to trigger the import/export dance all over the place :wink: